core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala

```diff
@@ -160,7 +160,13 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
           releaseLock(pieceId)
         case None =>
           bm.getRemoteBytes(pieceId) match {
-            case Some(b) =>
+            case Some(splitB) =>
+
+              // Checksum computation and further computations require the data
+              // from the ChunkedByteBuffer to be merged, so we we merge it now.
+              val bMergedData = splitB.toArray
```

Comment:
As I see when checksum is not enabled then merge is not needed.
And you can go further you can extend the checksum calculation to work on an array of chunks as I think it is just calling Alder32's `update` in a iteration for all the chunks.